### DISCUSSION TITLE
Read Array of data in get_handler for Autoevents

### DISCUSSION THREAD

**Discussion by vijay-hcl-epl** [edgexfoundry/community#176] [Q&A] [answered]
Hi all,
I am developing a device service in EdgeX 3.0 with CSDK 3.0.2. I am returning an array of integers in get_handler. So I am using iot_data_alloc_array for readings[i].value as shown below

`readings[i].value = iot_data_alloc_array (frame.data, frame.can_dlc, IOT_DATA_UINT8, IOT_DATA_COPY);`

However after autoevents, below error is shown in the core data logs:

> level=ERROR ts=2023-09-09T08:38:01.229513058Z app=core-data source=subscriber.go:77 msg="fail to unmarshal event, The value does not match the Uint8Array valueType"

Also the read data is not stored in the core data when I read the count or readings using curl core-data(59880) command. 
`curl -X 'GET' http://localhost:59880/api/v3/reading/resourceName/can_rdmsg | json_pp | more
`

If I read using CURL core command(59882), I am able to read the array of data properly.

`curl -X 'GET' 'http://localhost:59882/api/v3/device/name/candev/rdmsg' | json_pp`

Could you please let me know how to return a array of values in get_handler so that autoevents stores them in core data? Any sample code will be helpful

Thank you,
Vijay

**Comment by FelixTing** [comment]
The issue is caused by the inconsistency between the separator used by [iotech-c-utils](https://github.com/IOTechSystems/iotech-c-utils/tree/v1.5-branch) when converting array values to a JSON string and the one specified in [go-mod-core-contracts](https://github.com/edgexfoundry/go-mod-core-contracts/tree/main). go-mod-core-contracts uses ", "(comma followed by a space) as separator, while iotech-c-utils uses ","(comma without a space).

- go-mod-core-contracts: https://github.com/edgexfoundry/go-mod-core-contracts/blob/171c2b7d2fdecdd73b691d0ed80f4aaf6c80f917/dtos/reading.go#L189
- iotech-c-utils: https://github.com/IOTechSystems/iotech-c-utils/blob/60e1ab83fc9e87e86a1a7645e25c7995bbce2a35/src/c/data-json.c#L99

This will result in the reading value sent from the C Device Service to core-data failing to pass the validation of Reading DTO. https://github.com/edgexfoundry/go-mod-core-contracts/blob/171c2b7d2fdecdd73b691d0ed80f4aaf6c80f917/dtos/requests/event.go#L45
Because during the validation process, it splits the JSON string value into a string slice only by ", "(comma followed by a space). https://github.com/edgexfoundry/go-mod-core-contracts/blob/171c2b7d2fdecdd73b691d0ed80f4aaf6c80f917/dtos/reading.go#L410
```
func parseArrayValue(valueType string, value string) (err error) {
	arrayValue := strings.Split(value[1:len(value)-1], ", ") // trim "[" and "]"
        ...
```

In my opinion, we could modify the `parseArrayValue` function above to also use ","(comma without a space) as a separator.
@cloudxxx8 Do you see any potential impact?

**Reply by cloudxxx8** [reply]
@FelixTing yes, the validation should allow both ", " and ","
please make the change to use comma without a space as a separator, and trim the space for each element

**Reply by vijay-hcl-epl** [reply]
@FelixTing @cloudxxx8 , Thank you for the responses. I am currently running edgex as stand alone services. Is there any workaround I can use for now? Please let me know.

**Reply by cloudxxx8** [reply]
No, I don't think there is any workaround.

**Reply by vijay-hcl-epl** [reply]
Thank you @cloudxxx8 . Just to confirm, we need to wait until the next release of go-mod-core-contracts or CSDK for this to work right? Pls confirm

**Reply by cloudxxx8** [reply]
There will be no change from CSDK.
Are you using the development release v3.1 images?

**Reply by vijay-hcl-epl** [reply]
@cloudxxx8 ,
I am not using docker and running the edgex services individually. Latest edgex version is used. CSDK version is v3.0.2. I need a way to push the array of data from a C device service to core data. Please let me know how this can be done.

**Reply by FelixTing** [reply, marked as answer]
Hi @vijay-hcl-epl,
This issue has been fixed in go-mod-core-contracts v3.1.0-dev.14. Please modify the `go.mod` file in edgex-go to update the version of go-mod-core-contracts from v3.1.0-dev.13 to v3.1.0-dev.14, and then run `make data` to rebuild the core-data executable.

**Reply by vijay-hcl-epl** [reply]
Thank you @FelixTing.. It worked..!! Closing this issue..
